File: 2021/06/19/main.py
class Solution:
    def kInversePairs(self, n: int, k: int) -> int:
        m = 10 ** 9 + 7
        dp = [[0] * (k+1) for _ in range(n+1)]
        for i in range(n + 1): dp[i][0] = 1
        for i in range(1, n + 1):
            for j in range(1, k + 1):
                if j >= i:
                    dp[i][j] = dp[i][j-1] + dp[i-1][j] - dp[i-1][j-i]
                else:
                    dp[i][j] = dp[i][j-1] + dp[i-1][j]
                dp[i][j] %= m
        return dp[n][k]

File: 2021/06/19/main.go
package main

func kInversePairs(n int, k int) int {
	m, dp := 1000000007, make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, k+1)
		dp[i][0] = 1
	}
	for i := 1; i <= n; i++ {
		for j := 1; j <= k; j++ {
			if j >= i {
				dp[i][j] = dp[i][j-1] + dp[i-1][j] - dp[i-1][j-i]
			} else {
				dp[i][j] = dp[i][j-1] + dp[i-1][j]
			}
			dp[i][j] %= m
			if dp[i][j] < 0 {
				dp[i][j] += m
			}
		}
	}
	return dp[n][k]
}
